README
These are the release notes for  openmediavault-greyhole

openmediavault-greyhole is a plug-in for OpenMediaVault provinding the functionality in Greyhole. The software licensed under GPL 3+ with copyright ditributed among authors. See COPYING for details.

DOCUMENTATION
Please refer to the web-site http://www.openmediavault.org/ for pointers to documentation.

INSTALLING OPENMEDIAVAULT-GREYHOLE
If you want the latest stable openmediavault-greyhole plug-in, simply use the binary distribution. You will find Debian packages here: http://packages.omv-plugins.org/ providing a list of available plugins in the OpenMediaVault web-interface. However, if you want to test, install or develop the plugin, you should consider downloading and compiling from source.

DOWNLOADING AND BUILDING OPENMEDIAVAULT-GREYHOLE
For Debian based systems (such as OpenMediaVault itself), installing dependencies with:
 apt-get install git fakeroot debhelper build-essential
Downloading with:
 git clone https://github.com/OMV-Plugins/openmediavault-greyhole.git
 cd openmediavault-greyhole
 git checkout develop
For a first time build you can simply do:
 fakeroot debian/rules clean binary
For subsequent builds, you will need to update sources and clean up:
 git fetch
 rm -rf usr/share/openmediavault/locale
 rm -f debian/rules
 git checkout usr/share/openmediavault/locale
 git checkout debian/rules
 git pull origin develop
 git-dch --snapshot --auto --ignore-branch
 git-buildpackage --git-ignore-branch --git-builder='dpkg-buildpackage -T binary' --git-ignore-new


debian/changelog
openmediavault-greyhole (0.4) stable; urgency=low

  * Upgrade to xmlstarlet 1.3.1.

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 09 Sep 2012 10:27:26 +0200

openmediavault-greyhole (0.3.3) unstable; urgency=high

  * Github Issue Fix #34 - [Mount shares locally] Add cifs-utils package + Tuning
  * Github Issue FIx #35 - [Mount shares locally] Password field isn't hidden

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 10 Jun 2012 18:09:35 +0200

openmediavault-greyhole (0.3.2) unstable; urgency=high

  * Github Issue Fix #33 - [SMB Shares] Remove button isn't greyed when no line is selected
  * Github Issue Fix #32 - [Translation] Settings tab not translated
  * Github Issue Fix #31 - [Pool] Pool management error when using Replace option
  * Github Issue Fix #4 - Remove Samba Shares to the Greyhole Storage
  * Add script to mount samba shares locally for local use of greyhole
  * Update locale

 -- Marcel Beck <marcel.beck@mbeck.org>  Thu, 07 Jun 2012 19:47:48 +0200

openmediavault-greyhole (0.3.1) unstable; urgency=high

  * Various Fixes

 -- Marcel Beck <marcel.beck@mbeck.org>  Wed, 23 May 2012 18:23:42 +0200

openmediavault-greyhole (0.3) unstable; urgency=high

  * Github Issue Fix #21 - OMV 0.3 Port
  * Github Issue Fix #29 - Error when selecting log entry in OMV log tab

 -- Marcel Beck <marcel.beck@mbeck.org>  Fri, 20 Apr 2012 23:43:11 +0200

openmediavault-greyhole (0.2.2) stable; urgency=high

  * Github Issue #22 - Can't validate SMB dialog box when editing for sticky/trash options

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 01 Apr 2012 23:11:46 +0200

openmediavault-greyhole (0.2.1) stable; urgency=high

  * Fix Problem with SMB Shares

 -- Marcel Beck <marcel.beck@mbeck.org>  Mon, 26 Mar 2012 21:44:14 +0200

openmediavault-greyhole (0.2) stable; urgency=low

  * Github Issue #4 - Remove Samba Shares to the Greyhole Storage
  * Github Issue #6 - Remove a disk from Greyhole storage pool
  * Github Issue #7 - Replace/Failure of a disk in Greyhole storage pool (Storage Pool tab)
  * Github Issue #8 - Empty Greyhole trash (Share tab)
  * Github Issue #13 - Add DB Section within Settings Tab
  * Github Issue #14 - Add Greyhole Log in OMV Log Manager
  * Github Issue #15 - Catch OMV e-mail notification to notify greyhole fsck alerts
  * Github Issue #16 - Change Fsck dialog in pool tab + Add Fsck dialog in share tab
  * Github Issue #18 - Sticky files
  * Github Issue #19 - Trash per share

 -- Marcel Beck <marcel.beck@mbeck.org>  Sun, 18 Mar 2012 12:09:27 +0100

openmediavault-greyhole (0.1) stable; urgency=low

  * Initial public release

 -- Marcel Beck <marcel.beck@mbeck.org>  Mon, 13 Feb 2012 22:38:30 +0100

openmediavault-greyhole (0.0.1) stable; urgency=low

  * Internal release

 -- Stephane Bocquet <stephane_bocquet@hotmail.com>  Tue, 16 Mar 2011 00:00:00 +0100


var/www/openmediavault/js/omv/module/greyhole/admin/panel/pools.js
/**
 * vim: tabstop=4
 *
 * @license http://www.gnu.org/licenses/gpl.html GPL Version 3
 * @author Stephane Bocquet <stephane_bocquet@hotmail.com>
 * @author Marcel Beck <marcel.beck@mbeck.org>
 * @copyright Copyright (c) 2011 Stephane Bocquet
 * @copyright Copyright (c) 2011 Marcel Beck
 * @version $Id: greyhole.js 12 2011-11-07 18:52:10Z
 *          stephane_bocquet@hotmail.com $
 *
 * This file is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * This file is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this file. If not, see <http://www.gnu.org/licenses/>.
 */
// require("js/omv/NavigationPanel.js")
// require("js/omv/data/DataProxy.js")
// require("js/omv/FormPanelExt.js")
// require("js/omv/form/SharedFolderComboBox.js")
// require("js/omv/form/plugins/FieldInfo.js")

// require("js/omv/module/greyhole/panel/navigation.js")

// require("js/omv/module/greyhole/admin/dialog/pooldisk.js")
// require("js/omv/module/greyhole/admin/dialog/poolmngt.js")
// require("js/omv/module/greyhole/admin/dialog/fsck.js")

// require("js/omv/module/greyhole/util/Format.js")

Ext.ns("OMV.Module.Storage.Greyhole.Admin");

/**
 * @class OMV.Module.Storage.Greyhole.Admin.PoolsPanel
 * @derived OMV.grid.TBarGridPanel
 * Display list of configured filesystems.
 */
OMV.Module.Storage.Greyhole.Admin.PoolsPanel = function (config) {
	var initialConfig = {
		hidePagingToolbar:false,
		autoReload       :true,
		stateId          :"85f1cbf2-23d3-4960-a803-b7fc34d42235",
		colModel         :new Ext.grid.ColumnModel({
			columns:[
				{
					header   :_("Volume"),
					sortable :true,
					dataIndex:"volume",
					id       :"volume",
					width    :50
				},
				{
					header   :_("Label"),
					sortable :true,
					dataIndex:"label",
					id       :"label"
				},
				{
					header   :_("Filesystem"),
					sortable :true,
					dataIndex:"type",
					id       :"type",
					width    :50
				},
				{
					header   :_("Path"),
					sortable :true,
					dataIndex:"path",
					id       :"path",
					width    :200
				},
				{
					header   :_("Space"),
					sortable :true,
					dataIndex:"percent_space",
					id       :"percent_space",
					renderer :this.space_renderer,
					scope    :this
				},
				{
					header   :_("Trash"),
					sortable :true,
					dataIndex:"trash_size",
					id       :"trash_size",
					renderer :this.trash_renderer,
					scope    :this,
					width    :50
				},
				{
					header   :_("Min Free"),
					sortable :true,
					dataIndex:"min_free",
					id       :"min_free",
					renderer :this.min_free_renderer,
					scope    :this,
					width    :50
				}
			]
		})
	};
	Ext.apply(initialConfig, config);
	OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.constructor.call(this, initialConfig);
};
Ext.extend(OMV.Module.Storage.Greyhole.Admin.PoolsPanel, OMV.grid.TBarGridPanel, {
	initComponent:function () {
		this.store = new OMV.data.Store({
			autoLoad  :true,
			remoteSort:false,
			proxy     :new OMV.data.DataProxy({"service":"Greyhole", "method":"getPoolList"}),
			reader    :new Ext.data.JsonReader({
				idProperty   :"uuid",
				totalProperty:"total",
				root         :"data",
				fields       :[
					{ name:"uuid" },
					{ name:"volume" },
					{ name:"label" },
					{ name:"type" },
					{ name:"path" },
					{ name:"total_space" },
					{ name:"used_space" },
					{ name:"free_space" },
					{ name:"trash_size" },
					{ name:"potential_available_space" },
					{ name:"min_free" }
				]
			})
		});
		OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.initComponent.apply(this, arguments);
		// Register event handler
		// Reselect previous selected rows after the store has been
		// reloaded, e.g. to make sure toolbar is updated depending on
		// the latest row record values.
		this.getSelectionModel().previousSelections = [];
		this.store.on("beforeload", function (store, options) {
			var sm = this.getSelectionModel();
			var records = sm.getSelections();
			sm.previousSelections = [];
			Ext.each(records, function (record, index) {
				sm.previousSelections.push(record.get("uuid"));
			}, this);
		}, this);
		this.store.on("load", function (store, records, options) {
			var sm = this.getSelectionModel();
			var rows = [];
			if (Ext.isDefined(sm.previousSelections)) {
				for (var i = 0; i < sm.previousSelections.length; i++) {
					var index = store.findExact("uuid",
									sm.previousSelections[i]);
					if (index !== -1) {
						rows.push(index);
					}
				}
			}
			if (rows.length > 0) {
				sm.selectRows(rows);
			}
		}, this);
	},

	initToolbar:function () {
		var tbar = OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.initToolbar.apply(this);

		tbar.insert(3, {
			id     :this.getId() + "-poolmngt",
			xtype  :"button",
			text   :_("Pool management"),
			icon   :"images/greyhole-poolmngt.png",
			handler:this.cbpoolmngtBtnHdl,
			scope  :this
		});

		tbar.insert(4, {
			id     :this.getId() + "-balance",
			xtype  :"button",
			text   :_("Files balance"),
			icon   :"images/greyhole-balance.png",
			handler:this.cbbalanceBtnHdl,
			scope  :this
		});

		tbar.insert(5, {
			id     :this.getId() + "-fsck",
			xtype  :"button",
			text   :_("Files check"),
			icon   :"images/greyhole-fsck.png",
			handler:this.cbfsckBtnHdl,
			scope  :this
		});

		tbar.insert(6, {
			id     :this.getId() + "-unfsck",
			xtype  :"button",
			text   :_("Cancel all checks"),
			icon   :"images/greyhole-unfsck.png",
			handler:this.cbunfsckBtnHdl,
			scope  :this
		});

		tbar.insert(7, {
			id     :this.getId() + "-emptytrash",
			xtype  :"button",
			text   :_("Empty trash"),
			icon   :"images/greyhole-emptytrash.png",
			handler:this.cbemptytrashBtnHdl,
			scope  :this
		});

		return tbar;
	},

	cbAddBtnHdl:function () {
		var wnd = new OMV.Module.Storage.Greyhole.Admin.PoolDiskDialog({
			uuid     :OMV.UUID_UNDEFINED,
			listeners:{
				submit:function () {
					this.doReload();
				},
				scope :this
			}
		});
		wnd.show();
	},

	cbEditBtnHdl:function () {
		var selModel = this.getSelectionModel();
		var record = selModel.getSelected();
		var wnd = new OMV.Module.Storage.Greyhole.Admin.PoolDiskDialog({
			uuid     :record.get("uuid"),
			listeners:{
				submit:function () {
					this.doReload();
				},
				scope :this
			}
		});
		wnd.show();
	},

	/** POOL MANAGEMENT HANDLER */
	cbpoolmngtBtnHdl:function () {
		var wnd = new OMV.Module.Storage.Greyhole.Admin.PoolMngtDialog({
			listeners:{
				success:function (wnd, path, diskmngt) {
					this.dopoolmngt(path, diskmngt);
				},
				scope  :this
			}
		});
		wnd.show();
	},
	dopoolmngt      :function (path, diskmngt) {
		OMV.Ajax.request(this.cbpoolmngtLHdl, this, "Greyhole", "poolMngt",
						{
							path    :String(path),
							diskmngt:String(diskmngt)
						}
		);
	},
	cbpoolmngtLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /POOL MANAGEMENT HANDLER */

	/** BALANCE HANDLER */
	cbbalanceBtnHdl:function () {
		this.dobalance();
	},
	dobalance      :function () {
		OMV.Ajax.request(this.cbbalanceLHdl, this, "Greyhole", "balance", []);
	},
	cbbalanceLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /BALANCE HANDLER */

	/** FSCK HANDLER */
	cbfsckBtnHdl:function () {
		var wnd = new OMV.Module.Storage.Greyhole.Admin.FSCKDialog({
			listeners:{
				success:function (wnd, path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata) {
					this.dofsck(path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata);
				},
				scope  :this
			},
			type     :"pool"
		});
		wnd.show();
	},
	dofsck      :function (path, email_report, dont_walk_metadata_store, find_orphaned_files, checksums, delete_rphaned_metadata) {
		OMV.Ajax.request(this.cbfsckLHdl, this, "Greyhole", "fsck",
						{
							path                    :String(path),
							email_report            :Boolean(email_report),
							checksums               :Boolean(checksums),
							dont_walk_metadata_store:Boolean(dont_walk_metadata_store),
							find_orphaned_files     :Boolean(find_orphaned_files),
							delete_orphaned_metadata:Boolean(delete_rphaned_metadata)
						}
		);
	},
	cbfsckLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /FSCK HANDLER */

	/** CANCEL FSCK HANDLER */
	cbunfsckBtnHdl:function () {
		this.dounfsck();
	},
	dounfsck      :function () {
		OMV.Ajax.request(this.cbunfsckLHdl, this, "Greyhole", "unfsck", []);
	},
	cbunfsckLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /CANCEL FSCK HANDLER */

	/** EMPTY TRASH HANDLER */
	cbemptytrashBtnHdl:function () {
		this.doemptytrash();
	},
	doemptytrash      :function () {
		OMV.Ajax.request(this.cbemptytrashLHdl, this, "Greyhole", "emptyTrash", []);
	},
	cbemptytrashLHdl  :function (id, response, error) {
		if (error !== null) {
			// Display error message
			OMV.MessageBox.error(null, error);
		} else {
			OMV.MessageBox.hide();
			this.doReload();
		}
	},
	/** /EMPTY TRASH HANDLER */

	startDeletion     :function (model, records) {
		if (records.length <= 0)
			return;
		OMV.MessageBox.show({
			title  :_("Delete Pool Disk"),
			msg    :_("Do you want to remove the content of the pool disk directory recursively? Note, the data will be permanently deleted then. Select 'No' to delete the pool disk directory only or 'Cancel' to abort."),
			buttons:Ext.Msg.YESNOCANCEL,
			fn     :function (answer) {
				this.deleteRecursive = false;
				switch (answer) {
					case "yes":
						OMV.MessageBox.show({
							title  :_("Confirmation"),
							msg    :_("Do you really want to remove the pool disk directory content?"),
							buttons:OMV.Msg.YESCANCEL,
							fn     :function (answer) {
								if (answer === "yes") {
									this.deleteRecursive = true;
									OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.startDeletion.call(this, model, records);
								}
							},
							scope  :this,
							icon   :Ext.Msg.QUESTION
						});
						break;
					case "no":
						OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.startDeletion.call(this, model, records);
						break;
					case "cancel":
						break;
				}
			},
			scope  :this,
			icon   :Ext.Msg.QUESTION
		});
	},

	doDeletion:function (record) {
		OMV.Ajax.request(this.cbDeletionHdl, this, "Greyhole", "deletePoolDisk", { 'uuid': record.get("uuid"), 'recursive': this.deleteRecursive });
	},

	afterDeletion:function () {
		OMV.Module.Storage.Greyhole.Admin.PoolsPanel.superclass.afterDeletion.apply(this, arguments);
		delete this.deleteRecursive;
	},

	space_renderer:function (val, cell, record, row, col, store) {
		var total_space = parseInt(record.get("total_space")) * 1024;
		var used_space = parseInt(record.get("used_space")) * 1024;
		var free_space = parseInt(record.get("free_space")) * 1024;
		var percentage = parseFloat(used_space / total_space);

		if (-1 == total_space) {
			return val;
		}
		var id = Ext.id();
		(function () {
			new Ext.ProgressBar({
				renderTo:id,
				value   :percentage,
				text    :OMV.Module.Storage.Greyhole.Util.Format.bytesToSize(used_space) + '/' + OMV.Module.Storage.Greyhole.Util.Format.bytesToSize(total_space) + ' (' + parseInt(percentage * 100) + '%)'
			});
		}).defer(25);
		return '<div id="' + id + '"></div>';
	},

	min_free_renderer:function (val, cell, record, row, col, store) {
		val = val + ' GiB';
		return val;
	},

	trash_renderer:function (val, cell, record, row, col, store) {
		var trash = parseInt(val) * 1024;
		return OMV.Module.Storage.Greyhole.Util.Format.bytesToSize(trash);
	}
});

OMV.NavigationPanelMgr.registerPanel("storage", "greyhole", {
	cls     :OMV.Module.Storage.Greyhole.Admin.PoolsPanel,
	position:20,
	title   :_("Pool")
});
